covidapi/services/crud.py

```diff
@@ -12,8 +13,26 @@ def get_daily_report(self, db: Session, daily_report_id: int):
                 models.JHDailyReport.id == daily_report_id
             ).first()
 
-    def get_daily_reports(self, db: Session, skip: int = 0, limit: int = 100):
-        return db.query(models.JHDailyReport).offset(skip).limit(limit).all()
+    def get_daily_reports(
+        self, db: Session, skip: int = 0, limit: int = 100,
+        last_update_from: date = None,
+        last_update_to: date = None,
+        country: str = None,
+        province: str = None,
+    ):
+        query = db.query(models.JHDailyReport)
+
+        if last_update_from:
+            query = query.filter(models.JHDailyReport.last_update >= last_update_from)
+        if last_update_to:
+            query = query.filter(models.JHDailyReport.last_update <= last_update_to)
+        if country:
+            query = query.filter(models.JHDailyReport.country_region == country)
```

Comment:
This parameter could be interpreted in two ways:

- Give me data for this country as a whole
- Give me data for regions within this country

I noticed when I was writing the import script that some data has become more granular over time, so it would be possible for a response to return data covering a mix of geographical areas.

In the case of the US, you would get county level data (see https://github.com/CSSEGISandData/COVID-19/issues/1250).

I also noticed that the UK [originally meant the UK](https://github.com/CSSEGISandData/COVID-19/blob/master/csse_covid_19_data/csse_covid_19_daily_reports/01-31-2020.csv) but this then changed to a country/region of "United Kingdom" which has an entry for the UK, but also includes rows for overseas territories. See for example https://github.com/CSSEGISandData/COVID-19/blob/master/csse_covid_19_data/csse_covid_19_daily_reports/03-30-2020.csv

This seems like it would be quite difficult to consume and we should consider doing some more data cleaning in the import. I can't think of any case where it would be useful to query the UK and get data for Bermuda.

We could try and map these these to a standard country codes e.g. [ISO 3166](https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes). JHU have published a mapping we can use here https://github.com/CSSEGISandData/COVID-19/issues/1791. I don't think this will be perfect, because we have data points like the channel islands which technically covers multiple ISO 3166 country codes, but it will be better than relying on identifiers that change over time.

Reply:
> This parameter could be interpreted in two ways:
> 
>     * Give me data for this country as a whole
> 
>     * Give me data for regions within this country
> 
> 
> I noticed when I was writing the import script that some data has become more granular over time, so it would be possible for a response to return data covering a mix of geographical areas.
> 
> In the case of the US, you would get county level data (see [CSSEGISandData/COVID-19#1250](https://github.com/CSSEGISandData/COVID-19/issues/1250)).
> 
> I also noticed that the UK [originally meant the UK](https://github.com/CSSEGISandData/COVID-19/blob/master/csse_covid_19_data/csse_covid_19_daily_reports/01-31-2020.csv) but this then changed to a country/region of "United Kingdom" which has an entry for the UK, but also includes rows for overseas territories. See for example https://github.com/CSSEGISandData/COVID-19/blob/master/csse_covid_19_data/csse_covid_19_daily_reports/03-30-2020.csv
> 
> This seems like it would be quite difficult to consume and we should consider doing some more data cleaning in the import. I can't think of any case where it would be useful to query the UK and get data for Bermuda.
> 
> We could try and map these these to a standard country codes e.g. [ISO 3166](https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes). JHU have published a mapping we can use here [CSSEGISandData/COVID-19#1791](https://github.com/CSSEGISandData/COVID-19/issues/1791). I don't think this will be perfect, because we have data points like the channel islands which technically covers multiple ISO 3166 country codes, but it will be better than relying on identifiers that change over time.

Should I leave the name of the paramter consistend with the field name?
Looking at the data "country_region" has values like US, Italy, France etc... while province_state has names for US states (New York, California, Texas etc....).

Unfortunately JH CSSE confirms to be a very bad source of data, but there isn't much we can do about it. I hope we can support specific countries data sources as soon as possible.

I know that UK has released something similar to "Italian Protezione Civile". I think we should concentrate our efforts on those data sources, because the JH CSEE is only good to have a very generic "picture" of the various countries.

Reply:
Perhaps for this PR, we could leave it consistent with the field name, and then we can create another issue/PR to link to ISO or WHO country codes. That way we can add in extra fields/filters that are more reliable later on.

I think JH CSSE are doing their best (that table they released is definitely better than nothing) but a lot of the problems come from individual countries' reporting, and they are passing along the problems. This has definitely made me appreciate the value of a strict/well managed schema though, and I think our API can help with that, it's just time consuming and fiddly.

Regarding the UK data, ONS publish a dataset of [registered deaths](https://www.ons.gov.uk/peoplepopulationandcommunity/birthsdeathsandmarriages/deaths/datasets/weeklyprovisionalfiguresondeathsregisteredinenglandandwales) we could use. That data should be reliable, but I guess it will lag behind reality a bit.

Public Health England publish [case information](https://www.gov.uk/government/publications/covid-19-track-coronavirus-cases), which I haven't looked at yet but I assume it's where all the JH CSSE data is coming from the UK counts.

Besides active cases, recoveries and deaths we could also look at leading indicators like proactive testing, provision of PPE/ventilators, or public transport usage, but I don't know how much of that is available as open data in the UK.

Should we create issues for some of this?

Reply:
Btw I thought this was an interesting article about some of the problems with official counts and what metrics are reliable/predictive: https://pandemic.substack.com/p/the-2-coronavirus-metrics-you-should

Reply:
> Btw I thought this was an interesting article about some of the problems with official counts and what metrics are reliable/predictive: https://pandemic.substack.com/p/the-2-coronavirus-metrics-you-should

I've read the article, it begins correctly but misses the point entirely: the most useful metric at the moment is the number of new infection by the number of new tests (it's epidemiologists saying this, not my own conclusion).

It would be pointless to read "today there are +3000 new positive" without knowing how many tests they did. Italian Protezione Civile includes these numbers for example.

p.s: going back to the PR what should we do here? I'm not sure if/what I should fix 🤔

Thanks!

Reply:
ah sorry, I think let's merge it as is for now and do some work on country codes later.

Reply:
Thanks. I will update the README a little bit after merging this, to add a couple of examples and update the status of the project. Also the live API isn't linked anywhere yet, but I need to double check something before making the link public.